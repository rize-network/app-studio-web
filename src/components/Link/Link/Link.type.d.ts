// Defines a union type for text decoration styles which can be 'default', 'hover', or 'underline'.
export type TextDecorationStyle = 'default' | 'hover' | 'underline';
// Begins the declaration of a 'Styles' type that will specify the shape of style properties for the Link component.
export type Styles = {
// Optional 'icon' property of type 'CSSProperties' to define styles for the icon within the link.
  icon?: CSSProperties;
// Optional 'text' property of type 'CSSProperties' to define styles for the text within the link.
  text?: CSSProperties;
};
// Begins the declaration of a 'Size' type to dictate allowable size options for the Link component.
export type Size =
// The 'Size' type includes 'xs' as one of its possible values for the size of the Link component.
  | 'xs'
  | 'sm'
  | 'md'
  | 'lg'
  | 'xl'
  | '2xl'
  | '3xl'
  | '4xl'
  | '5xl'
// The 'Size' type includes '6xl' as its largest defined value for the size of the Link component.
  | '6xl';
