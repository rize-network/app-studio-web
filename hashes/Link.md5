3ffa878f499e22385f119e418b6eeb53