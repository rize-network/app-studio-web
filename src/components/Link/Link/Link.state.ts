import { useState } from 'react';
// This function initializes the Link component's state using the useState hook and exposes its state and setter function.
export const useLinkState = () => {
  const [isHovered, setIsHovered] = useState(false);
  return { isHovered, setIsHovered };
};
