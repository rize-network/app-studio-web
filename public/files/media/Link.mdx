# Link

Enables users to navigate to specified URLs or paths.

### **Import**
  ```tsx
  import { Link } from '@app-studio/web';
  ```

### **Default**
```tsx
import React from 'react';

import { Link } from '../Link';

export const DefaultLink = () => <Link href="/">Default</Link>;

```

### **underline**
Optional prop to define the text decoration style of the link as underline styles.

```tsx
import React from 'react';
import { Text } from '../../Text/Text';
import { Vertical } from '../../Layout/Vertical/Vertical';

import { Link } from '../Link';

export const UnderlineLink = () => (
  <Vertical gap={10}>
    <Link href={'https://www.npmjs.com/package/app-studio'} underline="default">
      Default
    </Link>
    <Link href={'https://www.npmjs.com/package/app-studio'} underline="hover">
      Hover
    </Link>
    <Link
      href={'https://www.npmjs.com/package/app-studio'}
      underline="underline"
      color="theme.primary"
      textDecorationColor="theme.primary"
    >
      <Text>Underline</Text>
    </Link>
  </Vertical>
);

```

### **isExternal**
Optional boolean indicating whether the link points to an external resource. Default behavior may vary based on this value.

```tsx
import React from 'react';
import { Text } from '../../Text/Text';

import { Link } from '../Link';

export const ExternalLink = () => (
  <Link
    href={'https://www.npmjs.com/package/app-studio'}
    isExternal
    iconSize="md"
  >
    <Text size="2xl">External</Text>
  </Link>
);

```

## Props

| Prop               | Type          | Description                                                                                                                               | Default |
| ------------------ | ------------- | ----------------------------------------------------------------------------------------------------------------------------------------- | ------- |
| href               | string        | The URL or path the link points to.                                                                                                       |         |
| underline          | string        | The text decoration style of the link ('default', 'hover', 'underline').                                                                 |         |
| isExternal         | boolean       | Whether the link points to an external resource.                                                                                          | false   |
| iconSize           | Size          | The size of the icon if the link is external ('xs', 'sm', 'md', 'lg', 'xl').                                                              | 'md'    |
| color              | string        | The color of the link text.                                                                                                               |         |
| textDecorationColor | string        | The color of the text decoration (e.g., underline).                                                                                       |         |
| styles             | CSSProperties | Optional custom styles for the link (container, text, icon).                                                                              |         |
| children           | React.ReactNode | The content of the link (typically text or an icon).                                                                                       |         |
