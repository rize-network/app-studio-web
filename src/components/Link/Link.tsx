import React from 'react';

import { LinkProps } from './Link/Link.props';
import { useLinkState } from './Link/Link.state';
import LinkView from './Link/Link.view';

const LinkComponent: React.FC<LinkProps> = (props) => {
  const linkStates = useLinkState();
  return <LinkView {...linkStates} {...props} />;
};

/**
 * Link allows users to navigate from page to page. It have a similar appearance as the hyperlink.
 */
export const Link = LinkComponent;
