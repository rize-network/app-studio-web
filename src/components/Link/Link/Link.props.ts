import React from 'react';
import { Size, Styles, TextDecorationStyle } from './Link.type';
import { ViewProps } from 'app-studio';
// Defines the props for the Link component, excluding 'size' which comes from the extended 'ViewProps'.
export interface LinkProps extends Omit<ViewProps, 'size'> {
  // Represents the content within the Link component which can be any valid React node.
  children: React.ReactNode;
  // Optional prop to define the text decoration style of the link as underline views.
  underline?: TextDecorationStyle;
  // Optional boolean indicating whether the link points to an external resource. Default behavior may vary based on this value.
  isExternal?: boolean;
  // Optional prop to set the size of an icon if present within the link.
  iconSize?: Size;
  // The href attribute specifying the URL the link points to.
  to: string;
  // Optional styles object to customize the appearance using predefined style types.
  views?: Styles;
}
// Extends the basic LinkProps with additional properties related to the Link view.
export interface LinkViewProps extends LinkProps {
  // Indicates whether the link is under a hover state, which may adjust the link's style.
  isHovered: boolean;
  // A function to update the hover state of the link.
  setIsHovered: Function;
}
