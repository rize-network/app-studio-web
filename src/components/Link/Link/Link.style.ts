import { ViewProps } from 'app-studio';
import { Size } from './Link.type';

// Defines a mapping of various size names (from 'xs' to '6xl') to their respective numeric icon sizes, providing a scalable way to reference icon sizes throughout the component.
export const IconSizes: Record<Size, number> = {
  xs: 12,
  sm: 14,
  md: 16,
  lg: 18,
  xl: 20,
};

/**
 * Link base styles matching shadcn/ui patterns
 */
export const LinkBaseStyles: ViewProps = {
  textDecoration: 'underline',
  textUnderlineOffset: '2px',
  textDecorationThickness: '1px',
  transition: 'color 0.15s ease, opacity 0.15s ease',
  _hover: {
    opacity: 0.8,
  },
  _focus: {
    outline: 'none',
    opacity: 0.8,
  },
  _focusVisible: {
    outline: 'none',
    boxShadow: '0 0 0 2px rgba(255, 255, 255, 1), 0 0 0 4px rgba(0, 0, 0, 0.1)',
    borderRadius: '2px',
  },
};
