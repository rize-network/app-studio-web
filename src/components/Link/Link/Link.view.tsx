import React from 'react';
import { Element } from 'app-studio';
import { Link as ReactRouterLink } from 'react-router-dom';
import { Horizontal } from 'src/components';
import { ExternalLinkSvg } from 'src/components/Svg';

import { LinkViewProps } from './Link.props';
import { IconSizes } from './Link.style';

const LinkView: React.FC<LinkViewProps> = ({
  children,
  href = '/',
  iconSize = 'sm',
  underline = 'default',
  isHovered = false,
  isExternal = false,
  colorScheme = '#0072F5',
  styles = { icon: {}, text: {} },
  setIsHovered = () => {},
  ...props
}) => {
  const handleHover = () => {
    if (underline === 'hover') setIsHovered(true);
  };
  return (
    <ReactRouterLink to={href} target={isExternal ? '_blank' : '_self'}>
      <Element
        color={colorScheme}
        onMouseEnter={handleHover}
        onMouseLeave={handleHover}
        textDecoration={isHovered || underline === 'underline' ? 'underline !important' : 'none'}
        {...styles.text}
        {...props}
      >
        <Horizontal gap={3} alignItems="center" wrap="nowrap">
          {children}
          {isExternal && <ExternalLinkSvg color={colorScheme} size={IconSizes[iconSize]} style={styles.icon} />}
        </Horizontal>
      </Element>
    </ReactRouterLink>
  );
};

export default LinkView;
