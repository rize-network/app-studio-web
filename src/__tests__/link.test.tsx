import React from 'react';
import { BrowserRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import { Link } from 'src/components/Link/Link';

import { cleanup, render, screen } from '@testing-library/react';

afterEach(() => {
  cleanup();
});

test('should render Link component without crashing', () => {
  render(
    <BrowserRouter>
      <Link href={'/'}>Default</Link>
    </BrowserRouter>
  );
});

test('should render Link component with correct text', () => {
  render(
    <BrowserRouter>
      <Link href="/" role="Link">
        Link
      </Link>
    </BrowserRouter>
  );
  const LinkElement = screen.getByRole('Link');
  expect(LinkElement).toBeInTheDocument();
  expect(LinkElement).toHaveTextContent('Link');
});

test('should render Link component with correct class name', () => {
  render(
    <BrowserRouter>
      <Link role="Link" href={'https://www.npmjs.com/package/app-studio'} className="link">
        External
      </Link>
    </BrowserRouter>
  );
  const LinkElement = screen.getByRole('Link');
  expect(LinkElement).toHaveClass('link');
});

test('Link to have text decoration underline on hover should match snapshot', () => {
  const tree = renderer
    .create(
      <BrowserRouter>
        <Link href={'https://www.npmjs.com/package/app-studio'} underline="hover">
          Link
        </Link>
      </BrowserRouter>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test('Link to open external link in a new tab should match snapshot', () => {
  const tree = renderer
    .create(
      <BrowserRouter>
        {' '}
        <Link href={'https://www.npmjs.com/package/app-studio'} isExternal iconSize="sm">
          External
        </Link>
      </BrowserRouter>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
