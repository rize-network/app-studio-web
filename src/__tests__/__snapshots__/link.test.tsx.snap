// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link to have text decoration underline on hover should match snapshot 1`] = `
<a
  href="https://www.npmjs.com/package/app-studio"
  target="_self"
>
  <div
    className="sc-bdfCDU fiYbZn"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <div
      className="sc-bdfCDU gtjvF"
    >
      Link
    </div>
  </div>
</a>
`;

exports[`Link to open external link in a new tab should match snapshot 1`] = `
Array [
  " ",
  <a
    href="https://www.npmjs.com/package/app-studio"
    target="_blank"
  >
    <div
      className="sc-bdfCDU fiYbZn"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <div
        className="sc-bdfCDU gtjvF"
      >
        External
        <div
          className="sc-bdfCDU kWVsuj"
        >
          <svg
            fill="white"
            height="14px"
            stroke="white"
            style={Object {}}
            viewBox="0 0 50 50"
            width="14px"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g
              id="SVGRepo_bgCarrier"
              strokeWidth="0"
            />
            <g
              id="SVGRepo_tracerCarrier"
              strokeLinecap="round"
              strokeLinejoin="round"
            />
            <g
              id="SVGRepo_iconCarrier"
            >
              <path
                d="M38.288 10.297l1.414 1.415-14.99 14.99-1.414-1.414z"
              />
              <path
                d="M40 20h-2v-8h-8v-2h10z"
              />
              <path
                d="M35 38H15c-1.7 0-3-1.3-3-3V15c0-1.7 1.3-3 3-3h11v2H15c-.6 0-1 .4-1 1v20c0 .6.4 1 1 1h20c.6 0 1-.4 1-1V24h2v11c0 1.7-1.3 3-3 3z"
              />
            </g>
          </svg>
        </div>
      </div>
    </div>
  </a>,
]
`;
